ban.go

```diff
@@ -0,0 +1,274 @@
+	"github.com/go-telegram-bot-api/telegram-bot-api"
+	"log"
+	"strings"
+	"sync"
+)
+
+const (
+	// File to store info on the requested bans.
+	requestedbansDB = "bans.json"
+	// Threshold to notify admins. If the offending message reaches this number
+	// of reports, the admins are notified.
+	adminNotificationDefaultThreshold = 1
+)
+
+// banRequest stores info on each ban requested via the /ban command.
+type banRequest struct {
+	// The ID of the offending message.
+	MessageID int64 `json:"message"`
+	// The ID of the group/channel in which the message has been reported.
+	ChatID int64 `json:"chat"`
+	// ID of the author of the message.
+	Author int64 `json:"author"`
+	// Text of the offending message, without considering any captions.
+	Text string `json:"text"`
+	// Map of users who reported this message, along with the ID of the report
+	// message.
+	Reporters map[int64]int64 `json:"reporters"`
+	// Map of notifications sent about this message. Here we have the ID of the
+	// admin notified and the ID of the notification (message) sent.
+	Notifications map[int64]int64 `json:"notifications"`
+	// Indicates whether this message has been removed by an admin.
+	MessageRemoved bool `json:"removed"`
+	// Admin who removed the message.
+	RemovedBy int64 `json:"handler"`
+}
+
+// banRequestList stores the list of bans requested alongside the threshold for
+// notifying the admins.
+type banRequestList struct {
+	// NotificationThreshold indicates the number of reports we need in order to
+	// notify the admins about a reported message.
+	NotificationThreshold int `json:"threshold"`
+	// Bans has as a key a combination of the ID of the reporting message with
+	// the ID of the group/channel this message happened. This is mapped to a
+	// banRequest struct, which contains additional info on the report itself.
+	Bans map[string]banRequest `json:"bans"`
+}
+
+type requestedBans struct {
+	sync.RWMutex
+	// List of requested bans alongside the threshold for notifying the admins.
+	Requests banRequestList
+}
+
+// banRequestHandler does it all.
+func (x *opBot) banRequestHandler(update tgbotapi.Update) error {
+	// This command is not supposed to be issued in private.
+	if update.Message == nil || update.Message.Chat == nil || isPrivateChat(update.Message.Chat) {
+		log.Printf("banRequestHandler: either i) the message or chat was nil or ii) this command was issued in a private message")
+		return nil
+	}
+
+	// If there is no message being replied to, we have no idea what exactly
+	// the user is reporting, so just ignore.
+	if update.Message.ReplyToMessage == nil {
+		log.Printf("banRequestHandler: /ban called by %q without being in response to a message; ignoring.", update.Message.From)
+		return nil
+	}
+
+	offendingMessageID := update.Message.ReplyToMessage.MessageID
+
+	x.reportedBans.Lock()
+	defer x.reportedBans.Unlock()
+
+	key := fmt.Sprintf("%d:%d", offendingMessageID, update.Message.Chat.ID)
+	_, ok := x.reportedBans.Requests.Bans[key]
```

Comment:
Could these two lines be in one, like if _, ok := x.reportedBans.Reqeusts.Bans[key]; ok { ... } ?